chapter_3/hello_lib/CMakeLists.txt

```diff
@@ -30,23 +25,45 @@ add_library(hello SHARED)
 # project version * SOVERSION set the compatibility  version for the library to
 # the major number of the version
 set_target_properties(
-  hello
-  PROPERTIES DEBUG_POSTFIX ${CMAKE_DEBUG_POSTFIX}
-             VERSION ${PROJECT_VERSION}
-             SOVERSION ${PROJECT_VERSION_MAJOR})
+    hello
+    PROPERTIES DEBUG_POSTFIX ${CMAKE_DEBUG_POSTFIX}
+               VERSION ${PROJECT_VERSION}
+               SOVERSION ${PROJECT_VERSION_MAJOR}
+)
 
 # add sources to the library target
 target_sources(
-  hello
-  PRIVATE src/hello.cpp src/internal.cpp
-  PUBLIC include/hello/hello.hpp)
+    hello
+    PRIVATE src/hello.cpp src/internal.cpp
+    PUBLIC include/hello/hello.hpp
```

Comment:
Header files do not need to be listed along with the source files. They do not get compiled.

The only purpose of listing them is when a developer is working on a project and when that developer is using an IDE that is "broken" and doesn't use headers from the include path to display.

Here we have 2 conditionals that must be met to warrant listing a header file:
* Developer mode
* IDE

Developer mode is something that is worth having a separate, but opt-in code path. However, it's pointless to check for the IDE, because for generators that aren't IDE specific listing the header file is a noop either way.

You can see me making use of this in [this example](https://github.com/friendlyanon/cmake-init-gif-engine/blob/master/cmake/dev-mode.cmake#L31) project. I had to make a concession there by using a custom `target_sources_grouped()` command, because it also does something that is only interesting to developers (defines source groups), but I wanted to avoid having a variable with a list of sources or reading out the properties of targets.

I also [mentioned](https://github.com/PacktPublishing/CMake-Tips-and-Tricks/issues/6#issuecomment-933491809) that I have not yet seen the usefulness of putting files in the `INTERFACE_SOURCES` property of a target, especially not a header file.

Reply:
You're right this can go in this case. 
Another use case I often encounter is when working with Qt, where the headers have to be parsed with the `moc` to generate signals and slots. According to the docs only the targets source files are processed when using `CMAKE_AUTOMOC`